Fix error message when using `const` inside a function, require the LHS to be global, and prohibit "const x[] = 1"

xal-0 opened the pull request:
The changes to `const` resulted in confusing error messages when it was used
inside a function (#57334).  On 1.11.3:

```
julia> function f()
           const x = 1
       end
ERROR: syntax: unsupported `const` declaration on local variable around REPL[1]:2
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```

On nightly:
```
julia> function f()
           const x = 1
       end
ERROR: syntax: World age increment not at top level
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```

In prior versions, we also accepted confused expressions like:
```
x = Ref(1)
const x[] = 1
```

This change adds a new error messages explicitly prohibiting `const` where the
left hand side is not declaring variables:
```
ERROR: syntax: `const` left hand side "x[]" contains non-variables around REPL[2]:1
Stacktrace:
 [1] top-level scope
   @ REPL[2]:1
```

Finally, #54773 made `const` stop participating in scope resolution (the left
hand side was always taken to be in global scope).  Some expressions that were
prohibited started being accepted:

In 1.11.3:
```
julia> let
           const x = 1
       end
ERROR: syntax: unsupported `const` declaration on local variable around REPL[1]:2
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```

Nightly:
```
julia> let
           const x = 1
       end
1
```

This change rejects `const` unless the variable would be in global scope
(`global const` would be required in the example), so we don't lose the
ability to make `const` in local scope meaningful later.

xal-0 committed:
Add tests for #57334, #57470

Also update some tests that used @test with a quote containing `const`, which
cannot be spliced into a function.

vtjnash reviewed (approved):
LGTM. Minor nit with this error though:
```
julia> const (x{T},) = T
ERROR: syntax: `const` left hand side "SSAValue(81)" contains non-variables around REPL[13]:1
```

xal-0 commented:
Argh, this is actually a regression.

1.11.3:
```
julia> const X{T}, Y{U} = [Int, Int]
2-element Vector{DataType}:
 Int64
 Int64
```

This:
```
julia> const X{T}, Y{U} = [Int, Int]
ERROR: syntax: `const` left hand side "SSAValue(4)" contains non-variables around REPL[1]:1
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```

I'll need to think of a way to handle this.

xal-0 committed:
Factor out expand-table '= lambda into expand-assignment

xal-0 committed:
const lowering: respect scope, prohibit non-const `const` assignment

Modifies the handling of the 3-argument "lowered const" so `const x = y`
participates in scope resolution again (#56613); we treat (const x y) as an
assignment form, producing the "unsupported `const` declaration on local
variable" error if the usual scope rules would not make x global.

At the top level, these should succeed:

    const x = 1
    begin
        const x = 1
    end
    let
        const global x = 1
    end

Each of the following should fail:

    # ERROR: syntax: unsupported `const` declaration on local variable around REPL[1]:2
    let
        const x = 1
    end

    # ERROR: syntax: unsupported `const` declaration on local variable around REPL[1]:2
    function f()
        const x = 1
    end

This also amkes the use of `const global` inside a function body produce the
original, specific error message rather than a generic "World age increment not
at top level" error (#57334):

    # ERROR: syntax: `global const` declaration not allowed inside function around REPL[1]:2
    function f()
        global const x = 1
    end

Issue #57334 revealed we silently accept `const` expressions writing to
variables.  These now fail with a new error message:

    a = [1, 2]
    # ERROR: syntax: cannot declare "a[1]" `const` around REPL[1]:2
    const a[1] = 3

The bulk of these changes track const-ness through the functions called by
expand-assignment, since each of the three types of LHS is affect differently:
- ordinary variables become `const`, producing error if not global
- `X{T} = ...` forces X to be `const`, rather than the defaulting to being
  implicitly const only if in global scope.
- `f() = ...` ignores `const`.

xal-0 commented:
This is a note to myself to not squash 9fe9bbeb5b557b7dabd0ec4043a5b87c0ef7ce15 so I am not git blame'd for all of the assignment lowering code.

xal-0 commented:
@nanosoldier `runtests()`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/57470#issuecomment-2679789484) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/ce76dbf_vs_095fc99/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**1 packages crashed only on the current version.**

 - An internal error was encountered: 1 packages

13 packages crashed on the previous version too.

## ✖ Packages that failed

**24 packages failed only on the current version.**

 - Package fails to precompile: 1 packages
 - Package has test failures: 1 packages
 - Package tests unexpectedly errored: 1 packages
 - Test duration exceeded the time limit: 21 packages

1109 packages failed on the previous version too.

## ✔ Packages that passed tests

**27 packages passed tests only on the current version.**

 - Other: 27 packages

5330 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**7 packages successfully loaded only on the current version.**

 - Other: 7 packages

2926 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

899 packages were skipped on the previous version too.



</details>

KristofferC commented:
Is https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/ce76dbf_vs_095fc99/PackageMaker.primary.log caused by this?

mlechu commented on src/julia-syntax.scm:
```diff
@@ -3126,7 +3152,7 @@
         ((eq? (car e) 'assign-const-if-global)
            (if (eq? (var-kind (cadr e) scope) 'local)
                (if (length= e 2) (null) `(= ,@(cdr e)))
```
```suggestion
               (if (length= e 2) (null) (resolve-scopes- `(= ,@(cdr e)) scope sp loc))
```

would fix #57574.

xal-0 replied:
Added!

xal-0 committed:
const lowering: resolve scopes under assign-const-if-global

Fixes #57574

Co-authored-by: Em Chu <im.mlechu@gmail.com>

vtjnash commented on test/syntax.jl:
```diff
@@ -3856,7 +3856,8 @@ const (gconst_assign(), hconst_assign()) = (2, 3)
 struct CantConvert; end
 Base.convert(::Type{CantConvert}, x) = error()
-@test (const _::CantConvert = 1) == 1
+# @test splices into a function, where const cannot appear
+@test Core.eval(@__MODULE__, :(const _::CantConvert = 1)) == 1
```
This seems misleading, since there are no functions created by `@test`, on master for example:
```
julia> let z = 2
       const z = 1
       z
       end
2

julia> z
1

julia> let z = 2
       global z = 1
       z
       end
ERROR: syntax: variable "z" declared both local and global
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1
```

Though I find this whole test highly unsatisfying since declaring `const _` is fairly surprising to mean no-op, while declaring `global _` is a side-effect-ful operation. Although not actually relevant to this PR since it is existing behaviors.
```
julia> Meta.@lower global _ = 1
:($(Expr(:thunk, CodeInfo(
1 ─     global _
│       $(Expr(:latestworld))
└──     return 1
))))

julia> names(Main)
4-element Vector{Symbol}:
 :Base
 :Core
 :Main
 :_
```

xal-0 replied:
Woops. I think I made that comment after causing the test to fail because of a bug, then I misinterpreted the output of `@macroexpand`.
